fgpyo/vcf/builder.py

```diff
@@ -0,0 +1,159 @@
+from pysam import VariantRecord as PysamVariant
+
+from fgpyo.vcf import Defaults as VariantBuilderDefaults
+from fgpyo.vcf import VariantBuilder
+from fgpyo.vcf import writer as PysamWriter
+
+
+class PysamVariantBuilder(VariantBuilder[PysamVariant]):
+    """
+    A variant builder for generating variants and VCF files, specifically using pysam to manage
+    writing the VCF.
+
+    Attributes:
+        samples: the sample name(s)
+        sd: list of dictionaries with names and lengths of each contig
+        seq_idx_lookup: dictionary mapping contig name to index of contig in sd
+        records: the list of variant records
+        header: the pysam header
+    """
+
+    samples: List[str]
+    sd: Dict[str, Dict[str, Any]]
+    seq_idx_lookup: Dict[str, int]
+    records: List[PysamVariant]
+    header: PysamHeader
+
+    def __init__(
+        self, samples: Optional[List[str]] = None, sd: Optional[Dict[str, Dict[str, Any]]] = None
+    ) -> None:
+        """Initializes a new VariantBuilder for generating variants and VCF files.
+
+        Args:
+            samples: the name of the sample(s)
+            sd: optional sequence dictionary
+        """
+        super().__init__(samples=samples, sd=sd)
+        self.header = PysamHeader()
+        for line in VariantBuilder.build_header_string(samples=samples, sd=self.sd).split("\n"):
+            if not line.startswith("##"):
+                break
+            self.header.add_line(line)
+        if samples is not None:
+            self.header.add_samples(samples)
+
+    def close(self) -> None:
+        """Closes this builder"""
+        pass
+
+    def add(
+        self,
+        contig: Optional[str] = None,
+        pos: Optional[int] = None,
+        id: Optional[str] = None,
+        ref: Optional[str] = None,
+        alts: Optional[Iterable[str]] = None,
+        qual: Optional[int] = None,
+        filter: Optional[Iterable[str]] = None,
+        info: Optional[Dict[str, Any]] = None,
+        samples: Optional[Dict[str, Dict[str, Any]]] = None,
+        input_start_index: int = 1,
+    ) -> PysamVariant:
+        """Generates a new variant and adds it to the internal collection.
+
+        Very little validation is done with respect to INFO and FORMAT keys being defined in the
+        header.
+
+        Args:
+            contig: the chromosome name
+            pos: the position
+            id: the variant id
+            ref: the reference allele
+            alts: the list of alternate alleles, None if no alternates
+            qual: the variant quality
+            filter: the list of filters, None if no filters (ex. PASS)
+            info: the dictionary of INFO key-value pairs
+            formats: the dictionary of sample name to FORMAT key-value pairs.
+            input_start_index: the genomic start index for input data (e.g. 0 for 0-based, 1-for
+                               1-based). VCF files are 1-based, but pysam is 0-based.
```

Comment:
I know pysam's SAM/BAM code is 0-based, but I think VCF is 1-based though I guess I have only tested .vcf and .vcf.gz files, and not .bcf files.

Reply:
After a little more poking I see the confusion.  `VariantRecord` has both `pos` which is 1-based like the VCF, but then also exposes attributes of `start` and `stop` (which are NOT vcf fields) which are 0-based open ended.  This is super confusing and honestly a really stupid decision in pysam to have.

Since I'd like to flatten this class into VariantBuilder, I think we can hide this somewhat and just accept 1-based positions and do this translation internally.

Reply:
Altered as per our conversation.